src/core/net/ip6.cpp

```diff
@@ -1492,6 +1505,41 @@ bool Ip6::IsOnLink(const Address &aAddress) const
     return rval;
 }
 
+#if OPENTHREAD_CONFIG_IP6_BR_COUNTERS_ENABLE
+void Ip6::UpdateBorderRoutingCounters(const Message &    aMessage,
+                                      const Header &     aHeader,
+                                      otPacketsAndBytes &aUnicastCounter,
+                                      otPacketsAndBytes &aMulticastCounter)
+{
+    otPacketsAndBytes *counter = nullptr;
+
+    if (!Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetSource()) &&
```

Comment:
Is all the checks correct/needed for both inbound and outbound? We seem to check here that neither src nor dest of message match any of unicast addresses of this device.

Suppose the device itself is sending a message to some device on infra link. 
- Then the src address of message will be one of the unicast addresses of this device itself
- We then filter it and not count it but it is actually going through border router. 

I think we can similar situation in opposite direction. Message we receive from infra link for this device itself will have one of the unicast addresses.

---
How about the following: 
- We can change the conditions based on whether the message is inbound or outbound.
- We can add a param to `UpdateBorderRoutingCounters()` to indicate if it is inbound or outbound (`bool` or `enum`)
- We can then determine which counters to update and also whether to check src or dest address.

Reply:
>Suppose the device itself is sending a message to some device on infra link.
>
> - Then the src address of message will be one of the unicast addresses of this device itself
> - We then filter it and not count it but it is actually going through border router.

In my current design, this is not defined as 'forwarded packet'. My target was to count all the border routed packets whose origin is not the BR itself. 
On the other hand, I wonder if this is a common use case. When will BR use its Thread source address to send a packet to infra link?

@simonlingoogle @librasungirl Should we update the counters' definition to include packets from BR itself?

Reply:
> In my current design, this is not defined as 'forwarded packet'. My target was to count all the border routed packets whose origin is not the BR itself.

I think it may makes sense to count them. 
- We want to count all Thread traffic (from any Thread node) that goes through the BR to infra network.
- The BR itself is also a Thread device and can generate traffic that goes through the BR.

Reply:
BR is a special device, agree to count if the traffic will go through both thread and infra interfaces, even if it's initiated by BR itself.

Reply:
I agree that it's a meaningful use case. I've updated the logic to count packets from/to the BR's Thread interface.

However, I just found that the current approach may not be able to count **all** such packets. 
For example, when a Wifi host sends a UDP packet to BR's Thread interface, the packet is handled by platform UDP instead of by Ip6 module in OpenThread core. Similarly, an outbound UDP packet sent by BR is also handled by platform UDP.

For the simplicity, may I focus on stats of the packets whose src/dest isn't the BR in this PR?

Reply:
Sounds ok to me.